src/lib/pubkey/hss_lms/lm_ots.h

```diff
@@ -0,0 +1,341 @@
+       */
+      static LMOTS_Params create_or_throw(LMOTS_Algorithm_Type type);
+
+      /**
+       * @brief Create the LM-OTS parameters from a hash function and width.
+       *
+       * @param hash_name tha name of the hash function to use.
+       * @param w the width (in bits) of the Winternitz coefficients.
+       * @throws Decoding_Error If the algorithm type is unknown
+       */
+      static LMOTS_Params create_or_throw(std::string_view hash_name, uint8_t w);
+
+      /**
+       * @brief Returns the LM-OTS algorithm type.
+       */
+      LMOTS_Algorithm_Type algorithm_type() const { return m_algorithm_type; }
+
+      /**
+       * @brief The number of bytes of the output of the hash function.
+       */
+      size_t n() const { return m_n; }
+
+      /**
+       * @brief The width (in bits) of the Winternitz coefficients.
+       */
+      uint8_t w() const { return m_w; }
+
+      /**
+       * @brief The maximum the winternitz coefficients can have.
+       */
+      uint8_t coef_max() const { return (1 << m_w) - 1; }
+
+      /**
+       * @brief The number of n-byte string elements that make up the LM-OTS signature.
+       */
+      uint16_t p() const { return m_p; }
+
+      /**
+       * @brief The number of left-shift bits used in the checksum function Cksm.
+       */
+      uint8_t ls() const { return m_ls; }
+
+      /**
+       * @brief Name of the hash function to use.
+       */
+      const std::string& hash_name() const { return m_hash_name; }
+
+   private:
+      /**
+       * @brief Construct a new LM-OTS parameter object.
+       *
+       * @param algorithm_type The algorithm type.
+       * @param hash_name The name of the hash function to use.
+       * @param w The width (in bits) of the Winternitz coefficients.
+       */
+      LMOTS_Params(LMOTS_Algorithm_Type algorithm_type, std::string_view hash_name, uint8_t w);
+
+      LMOTS_Algorithm_Type m_algorithm_type;
+      size_t m_n;
+      uint8_t m_w;
+      uint16_t m_p;
+      uint8_t m_ls;
+      std::string m_hash_name;
+};
+
+/**
+ * @brief Representation of a LM-OTS signature.
+ */
+class BOTAN_TEST_API LMOTS_Signature {
+   public:
+      /**
+       * @brief Parse a LM-OTS signature.
+       *
+       * @param slicer The private key bytes to parse.
+       * @return The LM-OTS signature.
+       * @throws Decoding_Error If parsing the signature fails.
+       */
+      static LMOTS_Signature from_bytes_or_throw(BufferSlicer& slicer);
+
+      /**
+       * @brief Returns the LM-OTS algorithm type.
+       */
+      LMOTS_Algorithm_Type algorithm_type() const { return m_algorithm_type; }
+
+      /**
+       * @brief The n-byte randomizer of the signature.
+       */
+      std::span<const uint8_t> C() const { return m_C; }
+
+      /**
+       * @brief Returns the part of the signature for @p chain_idx.
+       */
+      StrongSpan<const LMOTS_Node> y(uint16_t chain_idx) const { return m_y.at(chain_idx); }
```

Comment:
Sadly, I have no access to `params.n()` at this point to recompute the span. I could create a params object from m_algorithm_type, store n as a member, or add a function parameter for n. But none of these possibilities are very nice. I will keep it as it is since it's not expensive, and the buffer is never changed, so there is no danger.